aesd-char-driver/main.c

```diff
@@ -0,0 +1,202 @@
+#include "aesd-circular-buffer.h"
+int aesd_major =   0; // use dynamic major
+int aesd_minor =   0;
+
+MODULE_AUTHOR("Ashwin Ravindra"); /** TODO: fill in your name **/
+MODULE_LICENSE("Dual BSD/GPL");
+
+struct aesd_dev aesd_device;
+char* final_buffptr = NULL;
+
+int aesd_open(struct inode *inode, struct file *filp) {
+    PDEBUG("open");
+    struct aesd_dev *dev;
+    dev = container_of(inode->i_cdev, struct aesd_dev, cdev);
+    filp->private_data = dev;
+    return 0;
+}
+
+int aesd_release(struct inode *inode, struct file *filp) {
+    PDEBUG("release");
+    /*Nothing to handle here. aesd_cleanup_module is responsible for cleanup*/
+    return 0;
+}
+
+ssize_t aesd_read(struct file *filp, char __user *buf, size_t count, loff_t *f_pos) {
+    ssize_t retval = 0;
+    PDEBUG("read %zu bytes with offset %lld",count,*f_pos);
+    size_t* f_pos_offset = kmalloc(sizeof(size_t), GFP_KERNEL);
+    struct aesd_buffer_entry* read_entry;
+
+    /*Lock the buffer and find the entry to read*/
+    mutex_lock(&aesd_device.write_lock); 
+
+    /*If the entry is found, read the entry and copy to user*/
+    if((read_entry = aesd_circular_buffer_find_entry_offset_for_fpos(aesd_device.command_buffer, (size_t)*f_pos, f_pos_offset)) != NULL) {
+        ssize_t bytes_read = read_entry->size - *f_pos_offset;
+        if(bytes_read > count)
+            bytes_read = count;
+        copy_to_user(buf, read_entry->buffptr + *f_pos_offset, bytes_read);
+        *f_pos += bytes_read;
+        retval = bytes_read;
+    }
+
+    /*Unlock the buffer*/
+    mutex_unlock(&aesd_device.write_lock);
+    return retval;
+}
+
+ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count, loff_t *f_pos) {
+    PDEBUG("write %zu bytes with offset %lld",count,*f_pos);
+    ssize_t retval = -ENOMEM;
+    static size_t final_count = 0;
+
+    /*Temporary buffer to store the string from user*/
+    char* temp_buffptr = kmalloc(count, GFP_KERNEL);
+    copy_from_user(temp_buffptr, buf, count);
+
+    /*If newline character not found, append to final_buffptr*/
+    if(strchr(temp_buffptr, '\n') == NULL) {
+        final_buffptr = krealloc(final_buffptr, final_count + count, GFP_KERNEL); //realloc based on new size
+        memcpy(final_buffptr + final_count, temp_buffptr, count);
+        PDEBUG("final_buffptr: %s", final_buffptr);
+        final_count += count;
+        retval = count;
+        kfree(temp_buffptr); //this is not needed anymore
+        temp_buffptr = NULL; // Dont leave it dangling
+        return retval;
+    }
+
+    /*If newline character found, append to final_buffptr and write to buffer*/
+    final_buffptr = krealloc(final_buffptr, final_count + count, GFP_KERNEL);
+    memcpy(final_buffptr + final_count, temp_buffptr, count);
+    PDEBUG("final_buffptr: %s", final_buffptr);
+    final_count += count;
+
+    /*Write to the circular command buffer*/
+    struct aesd_buffer_entry *new_entry = kmalloc(sizeof(struct aesd_buffer_entry), GFP_KERNEL);
```

Comment:
@ashwinhr186 missing a free of this structure.  I'd recommend not dynamically allocating this.